src/layout/verticalPanel/allApplicationList.ts

```diff
@@ -0,0 +1,194 @@
+            } else if (symbol === Clutter.KEY_ISO_Left_Tab) {
+                this.selectPrevious();
+
+                return Clutter.EVENT_STOP;
+            } else if (symbol === Clutter.KEY_Down) {
+                this.selectNext();
+                return Clutter.EVENT_STOP;
+            } else if (symbol === Clutter.KEY_Up) {
+                this.selectPrevious();
+                return Clutter.EVENT_STOP;
+            } else if (
+                symbol === Clutter.KEY_Return ||
+                symbol === Clutter.KEY_KP_Enter
+            ) {
+                if (this.entrySelected !== null) {
+                    this.entrySelected.emit('primary-action');
+                }
+                return Clutter.EVENT_STOP;
+            }
+        }
+        return Clutter.EVENT_PROPAGATE;
+    }
+
+    updateResults() {
+        const appSystem = Shell.AppSystem.get_default();
+        const appsInstalled = appSystem
+            .get_installed()
+            .filter((appInfo) => {
+                try {
+                    const _ = appInfo.get_id(); // catch invalid file encodings
+                } catch (e) {
+                    return false;
+                }
+                return (
+                    appInfo.should_show() &&
+                    this.parentalControlsManager.shouldShowApp(appInfo)
+                );
+            })
+            .sort((a, b) =>
+                a.get_display_name().localeCompare(b.get_display_name())
+            );
+
+        for (let appInfo of appsInstalled) {
+            let icon = new St.Icon({
+                icon_size: 32,
+                gicon: appInfo.get_icon(),
+            });
+
+            const entry = new SearchResultEntry(
+                icon,
+                appInfo.get_display_name(),
+                // The remote search provider also provides a description field, but the app search does not
+                appInfo.get_description(),
+                true
+            );
+            entry.connect('primary-action', () => {
+                // It's important that we do this first because it will remove the focus grab that we use.
+                // This has the effect of restoring focus to the actor that was focused when we first grabbed it.
+                // So if we want to focus a newly created window we need to be sure to do it after we close the search view.
+                this.resetAndClose();
+
+                const app = Shell.AppSystem.get_default().lookup_app(
+                    appInfo.get_id()
+                );
+                if (app) {
+                    Me.msWindowManager.openApp(
+                        app,
+                        Me.msWorkspaceManager.getActiveMsWorkspace()
+                    );
+                }
+            });
+            this.addResult(entry);
+        }
+    }
+
+    addResult(resultEntry: SearchResultEntry): void {
+        this.add_child(resultEntry);
+        if (this.resultEntryList.length === 1) {
```

Comment:
Calling `resultEntryList` here means the `updateResults` function will become quadratic in the number of installed apps. Since it calls `addResult` N times (N = installed apps) and `resultEntryList` iterates through all children (`O(N)`) this will become `O(N^2)` which is kinda bad. I'd instead recommend that you call `selectResult` once, as the last thing you do in `updateResults`.

Reply:
Also. Is this initialized every time we open the list, or just once after the extension starts? If only once, then it may not discover new apps that are installed after the extension loads.